Host Registration - Install packages

stejskalleos opened the pull request:
New host macro `:manage_packages` for package installation
on supported OS families: Red Hat, Debian & Suse.

In host registration users can now install packages
on the host by setting the `host_registration_packages`.

stejskalleos committed:
Fixes #31743 - Template macro :manage_packages

For package installation on supported OS families: Red Hat, Debian & Suse.
In host registration users can now install packages
by setting the `host_registration_packages` parameter.

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,29 @@ def ks_console
           end
 
+          apipie :method, 'Returns manage_packages snippet with packages & action' do
+            required :action, String, "Allowed actions: 'install', 'update', 'remove', 'group install' and 'group remove'"
```
`apt update` and `yum update` are entirely different commands. The implementation suggests you have implemented things how yum does it, but as a user it would surprise me. `group` commands are not something `apt` knows either. This needs more documentation to describe what it does.

stejskalleos replied:
> apt update and yum update are entirely different commands. The implementation suggests you have implemented things how yum does it, but as a user it would surprise me.

Good point, this needs to be done differently.

> group commands are not something apt knows either. 

It's taken care of in the snippet.

> This needs more documentation to describe what it does.

Well the logic is in the snippet. This macro only returns the snippet, with some small logic to it.
If users wants to know what is happening, snippet is right place to go.

ekohl replied:
> Well the logic is in the snippet. This macro only returns the snippet, with some small logic to it.
> If users wants to know what is happening, snippet is right place to go.

Then the documentation should say that. Right now I think the documentation isn't very clear in that. I still question whether we should even have macros that wrap snippets in the first place. Until now we've always used the snippet tag inside templates when we wanted to reuse logic.

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,29 @@ def ks_console
 
+          apipie :method, 'Returns manage_packages snippet with packages & action' do
+            required :action, String, "Allowed actions: 'install', 'update', 'remove', 'group install' and 'group remove'"
+            required :packages, Array, 'Packages to manage'
```
It accepts strings too. Can apipie describe that?

stejskalleos replied:
> Can apipie describe that?

I checked docs but could not find how to define multiple types to one parameter.

ofedoren replied:
Welp, that's probably my bad :(

`apipie` for describing methods should almost fully compatible with `api` for describing endpoints.

Thus, if a parameter can accept string or array try to use
`required :packages, [Array, String], 'Packages to manage'`

ekohl commented on app/views/unattended/provisioning_templates/snippet/manage_packages.erb:
```diff
@@ -0,0 +1,90 @@
+  action = @action || 'install'
+  packages = @packages
+
+  if @host.operatingsystem.family == 'Redhat'
+    package_manager = 'yum'
```
Note that on Fedora there is no yum and it's just dnf. Fedora is part of the Redhat OS family. On EL8 there is also dnf but yum is provided for convenience.

stejskalleos replied:
But the yum is still there, right? I just set up fedora 31 image and can use yum without any problems.

ekohl replied:
Fedora 31 is EOL. There is a `yum` package and it does look like it's installed by default on Fedora 33 which surprises me.

adamruzicka replied:
Looks like it is just a symlink to dnf.

```shell
$ cat /etc/os-release | grep PRETTY_NAME                                                      
PRETTY_NAME="Fedora 33 (KDE Plasma)"

$ ls -la $(which yum)
lrwxrwxrwx. 1 root root 5 Dec  3 17:13 /usr/bin/yum -> dnf-3
```

While it may be there now, noone says it will still be there in future releases. I'd vote for not relying on it

stejskalleos commented:
@ekohl updated & rebased
Removed `manage_packages` macro, now just using `snippet_if_exists`.
Added `dnf` support for Fedora and dropped support for SUSE (Global Registration don't support it yet).

ares commented:
It seems we ship the snippet and only activate it if host param is defined. In such case, we should use `snippet` instead of `snippet_if_exists`. I think the idea with macro wasn't bad, it was the combination with the snippet that caused confusion. I think we should provide some reasonable way to install package (macro or snippet) and then, use it [here](https://github.com/theforeman/foreman_remote_execution/blob/master/app/views/templates/ssh/package_action.erb). That template has already more logic, e.g. for suse.

So I wonder, would it be better to take the content of the linked job template and convert it to macro in core? No snippet involved. The package list is based on the parameter. The only downside is, user can't extend it without Foreman code core change.

Also we could then use the same macro in our provisioning templates, where we today use `snippet_if_exists` logic, which was probably the motivation to do it the same way here.

ares commented:
Given the above, next steps:

1) @ekohl @stejskalleos do we prefer macro or snippet? I vote for macro
2) @stejskalleos would it be possible to base the implementation on the linked job template and modify that afterwards so it uses internally macro/snippet? I think that would payback long term

ekohl reviewed (commented):
I think that in principe a macro is a good idea, but I did struggle with the use of a snippet. We are providing an API with it. How clear is it to the user that if they make changes to that template, they can change the API?

One thing to consider: the commit message says "after host registration" but it calls `built` after installing packages. Perhaps `during` instead of `after` is a better description?

I would also consider solving this in a more generic way: don't include a parameter by default but give the user a generic `snippet_if_exists 'registration_linux_post'` so they can customize their registration via their own snippet.

ares commented:
Ok, so to summarize

1. let's add the macro, no snippet involved, based on the code in job template
2. let's use this macro to install packages specified in the parameter called e.g. `additional_packages` in linux registration template
3. ideally add the same thing to the kickstart templates
4. add <%= also snippef_if_exists 'registration_linux_post' %> to the linux registration template, so users can further customize the registration without modifying the default template
5. let's modify the rex job template to use that macro (that's rex repository work)

please speak up if I misunderstood

ekohl commented:
> ideally add the same thing to the kickstart templates

In kickstart you have 2 types of package installation: the native %packages section which also gives you a nice progress bar. However, it's generally not usable on RHEL due to subscriptions (though since RHEL 8.2 Anaconda understands activation keys so there it can be done, our kickstarts don't understand this).

I'd expect preseed to have something similar.

I'd be ok with leaving it out of scope for now.

Other than that :+1:

stejskalleos commented:
@ares @ekohl
Updated according to https://github.com/theforeman/foreman/pull/8298#issuecomment-774037850

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,41 @@ def ks_console
+            raise UndefinedPackages.new if packages.empty?
+            raise UnsupportedOS.new unless host.operatingsystem && ['Redhat', 'Debian', 'Suse'].include?(host.operatingsystem.family)
+
+            if host.operatingsystem.family == 'Redhat'
+              if host.operatingsystem.name.downcase == 'fedora'
+                return "dnf -y install #{packages}"
+              else
+                return "yum -y install #{packages}"
+              end
+            end
+
+            if host.operatingsystem.family == 'Debian'
+              return <<~CMD
+                [ -x "$(command -v subscription-manager)" ] && subscription-manager refresh
+                export DEBIAN_FRONTEND=noninteractive
+                apt-get -y update
+                apt-get -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" -y install #{packages}
+              CMD
+            end
+
+            if host.operatingsystem.family == 'Suse'
+              <<~CMD
+                zypper refresh
+                zypper -n install #{packages}
+              CMD
+            end
```
Mind making this a `case host.operatingsystem.family` statement instead of multiple if equal statements?

ares commented on app/views/unattended/provisioning_templates/registration/linux_registration_default.erb:
```diff
@@ -23,6 +23,8 @@ set -e
 <%= snippet 'insights' -%>
 
+<%= install_packages(host_param('host_registration_packages')) unless host_param('host_registration_packages').blank? %>
+<%= snippet_if_exists('registration_linux_post') %>
```
this will need renaming soon after the registration is renamed to host_init_config

ares commented on app/views/unattended/provisioning_templates/registration/linux_registration_default.erb:
```diff
@@ -23,6 +23,8 @@ set -e
 <%= snippet 'remote_execution_ssh_keys' %>
 <%= snippet 'insights' -%>
 
+<%= install_packages(host_param('host_registration_packages')) unless host_param('host_registration_packages').blank? %>
```
should we rename this to `host_init_packages` due to the upcoming rename of the template?

stejskalleos replied:
I'm not sure, all the other parameters are starting with `host_registration_*`,
I would say the relation to the registration process is more than to the template kind.

ares replied:
I think we said we want to use the same mechanism also in provisioning. Also for other params https://projects.theforeman.org/issues/32115

The "host init" seems to be applicable both in registration and provisioning processes

ares reviewed (changes requested):
Looks good, I'd just wait for the renaming PR to get in and rename things here so we don't have to rename again after merge.

stejskalleos commented:
Waiting for https://github.com/theforeman/foreman/pull/8256

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,39 @@ def ks_console
             (@port && @baud) ? "console=ttyS#{@port},#{@baud}" : ''
           end
 
+          apipie :method, 'Generates commands for package(s) installation, supported OS families: Redhat, Debian & Suse' do
```
This makes me wonder if there is a way to give a longer description. No sure if it exists, but would something like this make sense?
```suggestion
          apipie :method, 'Generates commands for package installation' do
            description 'Supported OS families: Redhat, Debian & Suse'
```

cc @ofedoren

ofedoren replied:
I've tried to add methods that could just came up in head of the person who writes the docs, so there are three methods you can use (aliases) to write longer description: `desc`, `description`, `full_description`.

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,39 @@ def ks_console
+            example "Fedora host: install_packages('pkg1 pkg2') #=> dnf -y install pkg1 pkg2"
+          end
+          def install_packages(packages)
+            raise UndefinedPackages.new if packages.empty?
```
Why is it needed to raise an exception for this? Can't it just return nil or an empty string? This would make it easier to use and I don't seen downsides.

ares replied:
Exception in macros are typically used to signal something unexpected happened and the exception message tells user what. I think the idea behind this was, if user calls the macro with empty list, it's probably not what they intended to do. However I agree with @ekohl that in this case, packages are mostly fed from the parameters which can be empty. Accepting (and ignoring) the empty list here would save users from having to add a condition in the template and checking for packages value presence.

While this is a general macro that can be also used without parameters, I think it's acceptable to not raise an error on empty string, it can return empty string in such case.

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,39 @@ def ks_console
+
+            case host.operatingsystem&.family
+            when 'Redhat'
+              if host.operatingsystem.name.downcase == 'fedora'
```
Technically on EL8+ DNF is also the package manager but yum is a symlink to dnf. I wonder if EL9 will still provide the compatibility symlink (wouldn't surprise me), but it's something to consider.

ares replied:
:+1: for using dnf on el8+ (ideally also for centos stream which we today recognize still as centos)

stejskalleos replied:
For both managers commands are same (`dnf/yum install -y pkg`), so not sure if there is need to have special condition for EL8+, or just keep it as it is.

ekohl replied:
The reasoning I'd have is that I'd want to be prepared for the future. If EL9 decides that `dnf` is the tool and drops `yum` then we won't have to make changes to support EL9. History has also taught us that a release is always closer than you think. It may be good to reach out internally to see what is the preferred tool.

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,39 @@ def ks_console
+              end
+            when 'Debian'
+              <<~CMD
+                [ -x "$(command -v subscription-manager)" ] && subscription-manager refresh
```
This is interesting. I never considered using `-x` on output of `command -v` but it does look like it returns non-executable files in `$PATH`. I learned something.

One thing I would consider here. If it runs in bash with `set -e` then I think it will exit if subscription-manager is not installed. If it's a real `if [ -x ...] ; then .. fi` block then you don't have this problem.

ares replied:
I guess this was added because subscription-manager can live in /usr/sbin/subscription-manager vs /usr/bin/subscription-manager

I think kickstart `set -e` did not work and registration does not use it, if it's a real issue, user would face it with rex templates. Given this was added in rex a [long time ago](https://github.com/theforeman/community-templates/commit/aeea6bca498e735894823f11bd14536997f6eb0c), I don't think it's broken and most likely we don't have to change it here.

ekohl replied:
> I guess this was added because subscription-manager can live in /usr/sbin/subscription-manager vs /usr/bin/subscription-manager

Not really. Normally `command -v` behaves like `which` but without printing to STDERR if it isn't found. It does also find shell built ins which `which` doesn't However, it looks like it can also find files in `$PATH` if they aren't executable. A bit of an edge case but a valid check.

stejskalleos replied:
I tried it and it won't exit if `subman` is not present:

```sh
#!/bin/sh

set -e
[ -x "$(command -v subscription-manager)" ] && subscription-manager refresh
echo "Say hi!"

# => Say hi!
```

ekohl replied:
Interesting, I didn't expect `set -e` to behave that way but it does look like it continues even with exit code 1. Looks like it continues regardless of anything when `&&` is used:
```console
$ cat bla
#!/bin/bash
set -e
[ -x "$(command -v false)" ] && false
echo $?
$ ./bla
1
$ cat bla2
#!/bin/bash
set -e
[ -x "$(command -v dfsfdfgdf)" ]
echo $?
$ ./bla2
$
```

So that means with `set -e`, `subscription-manager refresh` can fail and the script would still continue.

stejskalleos commented:
@ekohl @ares Rebased & updated

Added banner, Fedora & RHEL >=8 use dnf, fixed condition for subman on debian.

stejskalleos commented:
[test foreman]

ares commented on app/views/unattended/provisioning_templates/host_init_config/host_init_config_default.erb:
```diff
@@ -25,6 +25,9 @@ set -e
 
 <% end -%>
 
+<%= install_packages(host_param('host_packages')) unless host_param('host_packages').blank? -%>
```
we no longer need the `unless ...` part right?

ares commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,57 @@ def ks_console
+            case host.operatingsystem&.family
+            when 'Redhat'
+              fedora = host.operatingsystem.name.downcase == 'fedora'
+              rhel_dnf = host.operatingsystem.name.downcase == 'redhat' && host.operatingsystem.major.to_i >= 8
```
in fact the same applies to centos and probably future clones (alma, rocky, whatever), I guess we should use OS family == redhat && major > 8

the condition below would be fedora || redhat_family_dnf

ares reviewed (changes requested):
getting closer and closer :-) two small things and I'll test the whole thing

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,51 @@ def ks_console
+            when 'Redhat'
+              <<~CMD
+                #{banner}
+                if [ -f /usr/bin/dnf ]; then
```
I still think we should use host OS properties (from our models) to determine this rather than at runtime. From our Jenkins nodes running RHEL 7:
```console
$ rpm -qf /usr/bin/dnf
dnf-4.0.9.2-2.el7_9.noarch
$ cat /etc/os-release 
NAME="Red Hat Enterprise Linux Server"
VERSION="7.9 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.9"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.9 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.9:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.9
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.9"
```

However, we generally don't use dnf to install things. It's there to allow some RPM actions in our packaging workflows.

(Yes, this is all very messy)

stejskalleos replied:
Yeah but there is so many cases that this is the easiest solution to catch all of them. Simple and effective.

> However, we generally don't use dnf to install things. 

But in previous reviews you wanted to use `dnf`. Why is it bad now?

ekohl replied:
DNF on EL8 is good, DNF on EL7 is possible, but not what I would expect to be used. That's my concern. So what was mentioned in https://github.com/theforeman/foreman/pull/8298#discussion_r609032259 sounds close to what I would expect. I would expect (written in a shortened form):
```
if (is_fedora && version >= 22) || (!is_fedora && version >= 8)
```

stejskalleos replied:
But the problem here are the future clones of CentOS (Rocky & others), where the condition `(!is_fedora && version >= 8)` can be in future true but they won't have `dnf`, just `yum`. 

Anyway, here is a table with the options, is this correct?

OS | version | package manager
------------ | -------------| -------------
Fedora | all versions | `dnf`
CentOS |all|`yum`
RHEL|<= 7| `yum`
RHEL|>= 8|`dnf`
CentOS clones|all|`yum`

ekohl replied:
No, that's not correct. I've been too long in the OS world that I have to stop and remember this isn't just common knowledge to everyone. Here's the background.

Fedora started to use DNF in version 22, prior to that it was yum.

CentOS Linux is a rebuild of RHEL. For almost all intents and purposes, you can treat it as the same OS. That also means CentOS Linux is by definition behind RHEL (or at most, in sync). CentOS Linux has a bug-for-bug compatible with RHEL policy. This means that if a bug is reported and it's also present in RHEL, it's considered a WONTFIX and they tell you to get it fixed in RHEL instead. Fixing the bug would introduce different behavior from RHEL and is considered a bug. Note that CentOS Linux 8 will be EOL at the end of the year while 7 will remain supported until 2024 when regular RHEL support ceases.

Then we have the new kid CentOS Stream which is being positioned as an upstream of RHEL. This means it's similar, but it can actually be ahead of RHEL. Another big difference is that CentOS Stream does **not** have a minor version. There is just 8, not 8.x. This also means we will see a CentOS Stream 9 before RHEL 9.

Then there are other rebuilds of RHEL (not CentOS) like Rocky and AlmaLinux. They, like CentOS Linux, aim to be pure RHEL rebuilds. This is why you see the concept of OS Family: we can treat the Red Hat OS family (with the exception of Fedora) the same. This is also why you see EL used: it's RHEL and all its rebuilds.

Because of this, you see patterns in Puppet like
```puppet
case $facts['os']['family'] {
  'RedHat': {
    case $facts['os']['name'] {
      'Fedora': {
        # Fedora specific code
      }
      default: {
        if versioncmp($facts['os']['release']['major'], '8') >= {
          # EL8+
        } else {
          # EL7 and older
        }
      }
    }
  }
  'Debian': {
    case $facts['os']['name'] {
      'Ubuntu': {}
      default: {}
    }
  }
}
```

Foreman has its roots in Puppet and imported facts so it copied the OS family concept. That makes you understand the OS model in our database.

Then there are RHEL derivatives like OracleLinux and AmazonLinux. They are based on RHEL, but unlike rebuilds they aren't quite the same. Some base packages have changed or may not be shipped. OracleLinux is still quite close to RHEL while AmazonLinux has made much bigger changes. Facter no longer considers AmazonLinux part of the Red Hat OS family because of that while OracleLinux is. Due to all of this we consider them best effort supported. It is up to users of those distros to send fixes if there are problems and most developers never touch them. 

So in conclusion, it becomes:

OS | version | package manager
------------ | -------------| -------------
Fedora | >= 22 | `dnf`
Fedora | < 22 | `yum`
EL|>= 8|`dnf`
EL|< 8| `yum`

Does that help?

stejskalleos replied:
> Does that help?

Yes, a lot, thanks :) I'll update the PR according to your table.

stejskalleos commented:
[test katello]

stejskalleos commented:
[test katello]

ekohl commented on lib/foreman/renderer/scope/macros/host_template.rb:
```diff
@@ -121,6 +121,62 @@ def ks_console
+            when 'Redhat'
+              os = host.operatingsystem
+              is_fedora = os.name.downcase == 'fedora'
+              is_rhel = os.name.downcase == 'redhat'
+
+              if is_fedora
+                return <<~CMD
+                  #{banner}
+                  #{os.major.to_i >= 22 ? 'dnf' : 'yum'} -y install #{packages}
+                CMD
+              end
+
+              if is_rhel
+                return <<~CMD
+                  #{banner}
+                  #{os.major.to_i >= 8 ? 'dnf' : 'yum'} -y install #{packages}
+                CMD
+              end
+
+              "yum -y install #{packages}"
```
Based on my comment I would not have an explicit test for RHEL. Rather:
```suggestion
           is_dnf = (is_fedora && os.major.to_i >= 22) || (!is_fedora && os.major.to_i >= 8)
           <<~CMD
             #{banner}
             #{is_dnf ? 'dnf' : 'yum'} -y install #{packages}
           CMD
```
```

ares reviewed (approved):
I have tested the functionality and all works fine including the snippet. Leaving this open for a few days or until other reviews confirm concerns are addressed. If I don't see any reaction I plan to merge this by the end of the week.

ares commented:
Thanks @stejskalleos, merged and updated the commit message to reflect the new names
